mcs/class/corlib/Mono/SafeGPtrArrayHandle.cs

```diff
@@ -0,0 +1,42 @@
+//
+// Safe handle class for Mono.RuntimeGPtrArrayHandle
+//
+// Authors:
+//   Aleksey Kliger <aleksey@xamarin.com>
+//   Rodrigo Kumpera <kumpera@xamarin.com>
+//
+// Copyright 2016 Dot net foundation.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Mono {
+	internal sealed class SafeGPtrArrayHandle : IDisposable {
+		RuntimeGPtrArrayHandle handle;
+		bool freeSeg;
+
+		internal SafeGPtrArrayHandle (IntPtr ptr, bool freeSeg)
+		{
+			handle = new RuntimeGPtrArrayHandle (ptr);
+			this.freeSeg = freeSeg;
+		}
+
+		public void Dispose ()
+		{
+			RuntimeGPtrArrayHandle.DestroyAndFree (ref handle, freeSeg);
+			GC.SuppressFinalize (this);
```

Comment:
DestroyAndFree mutates handle to be null.

I think this does need a finalizer: in general if it ever gets used outside of a `using` it will leak unmanaged memory.